### PR TITLE
Add direnv and formatter for nix and shell

### PR DESCRIPTION
**Pull request by jfroche**
- Add direnv and formatter for nix and shell
- Fix formatting
- Upgrade to the latest nixpkgs stable

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -1,10 +1,31 @@
...
+            shellcheck.enable = true;
+            shfmt.enable = true;
+            statix = {
+              enable = true;
+              disabled-lints = [
+                "empty_pattern"
+                "eta_reduction"
+                "faster_groupby"
+                "faster_zipattrswith"
+              ];
```
Comment (soenkehahn) [inline flake.nix:23]: Why are these disabled? These lint rules don't seem to be triggering any warnings.

File: flake.nix [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
+  outputs = { self, nixpkgs, flake-utils, treefmt-nix }:
     flake-utils.lib.eachSystem [ "x86_64-linux" ] (system:
-      let pkgs = nixpkgs.legacyPackages.${system}; in
+      let
+        pkgs = nixpkgs.legacyPackages.${system};
+        treefmt-config = {
+          projectRootFile = "flake.nix";
+          programs = {
+            nixpkgs-fmt.enable = true;
+            shellcheck.enable = true;
```
Comment (soenkehahn) [inline flake.nix:14]: We already have a check running `shellcheck`, so I think now we're running it twice. We should remove one check.

File: flake.nix [hunk 3]
```diff
@@ -1,10 +1,31 @@
 {
   description = "Install scripts for nix-based tools";
   inputs.flake-utils.url = "github:numtide/flake-utils";
-  inputs.nixpkgs.url = "github:nixos/nixpkgs/23.05";
-  outputs = { self, nixpkgs, flake-utils }:
+  inputs.nixpkgs.url = "github:nixos/nixpkgs/23.11";
```
Comment (soenkehahn) [inline flake.nix:4]: Can we put the update to 22.11 into a different PR, please? If we care enough about updating, that is.

**Review by soenkehahn — CHANGES_REQUESTED** [review]
Overall this seems to make the flake file a lot more complicated for fairly little practical gain. Currently the only formatter that actually surfaces any issues is `shfmt`, I think. So I don't think I fully understand the motivation for this change.

Also, I find `treefmt` a bit opaque and underdocumented. Overall I think using it is not a terrible idea, but I would prefer to use something that is a little bit more accessible.

Also, this PR does multiple things at once and could be split up to discuss them separately. For example adding the `.envrc` file could be merged without further discussion.

File: flake.nix (hunk 1 above)
Comment (jfroche) [reply flake.nix:23]: Fixed in https://github.com/garnix-io/nix-tool-installer/pull/14

File: flake.nix (hunk 2 above)
Comment (jfroche) [reply flake.nix:14]: Fixed in https://github.com/garnix-io/nix-tool-installer/pull/14

File: flake.nix (hunk 3 above)
Comment (jfroche) [reply flake.nix:4]: Done here: https://github.com/garnix-io/nix-tool-installer/pull/13

**Comment by jfroche** [comment]
PR has been split

**Pull request closed without merging**